Item 9584: ManageSampleStatusesPanel for sample statuses CRUD within a given container

cnathe opened the pull request:
#### Rationale
Recent stories have added a Sample Status column and support for status change checks for permitted operations. This PR adds the UI component to manage the sample statuses available for a given container. It will be shown in LKS, LKSM, and LKB. In addition, this story adds in a SampleStatusService registered from the SM module so that by default the sample status support is not "enabled" unless the SM module is included in the build and registers the service implementation to support it.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/653
* https://github.com/LabKey/platform/pull/2698
* https://github.com/LabKey/sampleManagement/pull/726
* https://github.com/LabKey/biologics/pull/1026
* https://github.com/LabKey/inventory/pull/323

#### Changes
* remove ExpSchema and SampleStateManager isOperationPermitted() check in favor of calls to SampleStatusService method
* SampleStatusService and default provider for community distribution scenario
* AbstractManageQCStateAction validation checks for empty label value
* DataStatesTableInfo updates for QueryUpdateService validation checks for valid state change and empty label value
* SampleTypeWebPart update to conditionally add "Manage Sample Statuses" button to grid header when supportsSampleStatus() is true

cnathe committed:
Item 9584: SampleQCStateService
- default implementation allows for all sample operations regardless of status
- SM module implementation adds in the per status sample operation checks
- move stateType.operationPermitted from SampleStateType enum to service provider

cnathe committed:
rename SampleQCStateService to SampleStatusService
- and rename supportsQC method to supportsSampleStatus

cnathe committed:
Item 9584: ensure default sample states for a LKB or LKSM container
- on container creation of folder type Biologics or Sample Manager
- or when a folder type is changed to one of those two types
- only add the default states if the container does not have any sample states defined

cnathe committed:
Merge branch 'develop' into fb_smManageStatuses

cnathe committed:
Sample status QCStateHandler to check for in-use states and allowable state type changes

cnathe committed:
DataStatesTableInfo query update service checks for allowable state changes
- and Add checks for QC state label to not allow blanks

cnathe committed:
Update sample status related validation check error messages

cnathe committed:
ManageSampleStatusesPanel implementation for experiment module (LKS) and button/link on sample listing grid to get there, for admins

cnathe committed:
Merge branch 'develop' into fb_smManageStatuses

cnathe committed:
update @labkey/components package version

cnathe committed:
Only add sample type listing grid "Manage Sample Statuses" button when SampleStatusService supports status

cnathe committed:
Hide DataStates table's StateType column from insert/update views if sample status not supported

cnathe committed:
Move ManageSampleStatuses entryPoint from experiment module to sampleManagent module

cnathe committed:
DataStates StateType column updates to hidden and details view settings

cnathe committed:
update @labkey/components package version

cnathe committed:
in-use -> in use

cnathe committed:
Merge branch 'develop' into fb_smManageStatuses

# Conflicts:
#	experiment/src/org/labkey/experiment/api/ExpMaterialTableImpl.java

cnathe committed:
update @labkey/components package version

cnathe committed:
update @labkey/components package version

cnathe committed:
Export SampleState (i.e. status) as label instead of rowId for folder archive

labkey-susanh commented on api/src/org/labkey/api/qc/SampleStatusService.java:
```diff
@@ -0,0 +1,111 @@
+ */
+public interface SampleStatusService
+{
+    List<SampleStatusService> _providers = new ArrayList<>();
```
Why is this a list if you can only ever register one?

cnathe replied:
Two reasons:
1) I was following the pattern from another service interface
2) If I used a SampleStatusService variable and tried to set it in the registerProvider method, I get the following: `Cannot assign a value to final variable`

labkey-susanh commented on api/src/org/labkey/api/qc/SampleStatusService.java:
```diff
@@ -0,0 +1,111 @@
+        if (_providers.isEmpty())
+            _providers.add(provider);
+        else
+            throw new RuntimeException("An implementation of SampleStatusService has previously been registered");
```
```suggestion
            throw new RuntimeException("An implementation of SampleStatusService has previously been registered.");
```

cnathe replied:
Done

labkey-susanh commented on api/src/org/labkey/api/qc/DataStateHandler.java:
```diff
@@ -15,16 +15,28 @@
+     * @param rowUpdates The map of row changes for this state
+     * @return Error message to show the user if the change is not allowed
+     */
+    @Nullable String isStateChangeAllowed(Container container, DataState state, Map<String, Object> rowUpdates);
```
From the name I would expect this to return a boolean. Perhaps `getStateChangeError(...)`?

cnathe replied:
Good suggestion. Done.

labkey-susanh commented on core/src/org/labkey/core/query/DataStatesTableInfo.java:
```diff
@@ -80,9 +91,36 @@ private boolean validateQCStateNotInUse(Map<String, Object> oldRowMap, Container
+        private boolean validateLabel(Map<String, Object> row)
+        {
+            String label = (String) row.get("label");
+            return !StringUtils.isBlank(label);
```
Should we have a `trim()` in here?

cnathe replied:
Not here, it looks like the StringUtils.isBlank check already handles that. But I have added a trim() on the client side before calling to insert/update rows.

labkey-susanh commented on core/src/org/labkey/core/query/DataStatesTableInfo.java:
```diff
@@ -80,9 +91,36 @@ private boolean validateQCStateNotInUse(Map<String, Object> oldRowMap, Container
         protected Map<String, Object> updateRow(User user, Container container, Map<String, Object> row, @NotNull Map<String, Object> oldRow, boolean allowOwner, boolean retainCreation) throws InvalidKeyException, ValidationException, QueryUpdateServiceException, SQLException
         {
+            if (!validateLabel(row))
+                throw new QueryUpdateServiceException("State label cannot be blank.");
```
```suggestion
                throw new QueryUpdateServiceException("Label cannot be blank.");
```
This avoids using "State" when the user sees nothing but "status".

cnathe replied:
Done.

labkey-susanh commented on core/src/org/labkey/core/query/DataStatesTableInfo.java:
```diff
@@ -96,6 +134,9 @@ private boolean validateQCStateNotInUse(Map<String, Object> oldRowMap, Container
         protected Map<String, Object> insertRow(User user, Container container, Map<String, Object> row) throws DuplicateKeyException, ValidationException, QueryUpdateServiceException, SQLException
         {
+            if (!validateLabel(row))
+                throw new QueryUpdateServiceException("State label cannot be blank.");
```
```suggestion
                throw new QueryUpdateServiceException("Label cannot be blank.");
```

cnathe replied:
Done.

cnathe committed:
update @labkey/components package version

cnathe committed:
CR feedback
- rename SampleQCStateHandler to SampleStatusHandler
- rename isStateChangeAllowed() to getStateChangeError()
- add check in ensureDefaultStates() that the container is not of type "labbook"
- update "containersWithModuleCount" usage metric SQL to use prop.PropertyEntries

cnathe committed:
update @labkey/components package version

cnathe committed:
Sample timeline event include status label value so that it can still be shown if the status is removed

cnathe committed:
Attempt 2 to add Status Label to sample timeline event audit diff details

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_smManageStatuses

cnathe committed:
update @labkey/components package version
